metabase/metabase
# Normalize virtual card table fields using its `table_id` and `id`

Fixes https://github.com/metabase/metabase/issues/25141

`normalizr` docs: https://github.com/paularmstrong/normalizr/blob/master/docs/api.md

- I'm using the `idAttribute` config function on normalizr entities to create a unique `${table_id}:${id}` value for fields coming from virtual card tables.
- I'm using the `processStrategy` config function to add this unique identifier to the field object. I've named it `uniqueId`. This lets us avoid messing with the field's `id` property, which we frequently use directly to create field refs that go into mbql queries.
- I've updated the metadata selector code to use the `uniqueId` property for fields so that we avoid clobbering fields that share the same `id` value.
- I've updated the `Metadata` class `field` method to accept an optional `tableId` arg. Doing it this way helps avoid breaking existing code that calls `metadata.field(fieldId)` -- that'll work the same way. There are also scenarios (such as in admin data model code) where we don't need to use the second arg, anyways.
- I've updated one use of `metadata.field(fieldId)` to `metadata.field(fieldId, tableId)`, in some models code. There are other places where we use `metadata.field(fieldId)`, but I'm not totally certain which should be updated, so I'm leaving them alone for now. A small risk to this approach is that we fail to find any field where previously we would have. A fix might be to always fetch the base table of a Question, using the `card.table_id` property (I'm doing this anyway in https://github.com/metabase/metabase/pull/25109)
  - I'm also updating FieldDimension's usage of this function in PR #25109

Note that this doesn't fix all places where we are clobbering metadata because this PR does not update `FieldDimension.prototype.field`. I'm leaving that for https://github.com/metabase/metabase/pull/25109

**Testing**
This affects a lot of stuff. Try playing with...
- Create a question & click on column headers & mutate the underlying query (filters, drilldowns, etc.)
- Create a model & update the metadata of the model
- Create a question based on a model and see if its overrided metadata persists
- Add these questions to a dashboard and map field filter parameters to the fields and make sure things still work (ignore clobbering here)
- Make sure #22715 can't be repro'd 
- Same for https://github.com/metabase/metabase/issues/20419
